enhancements/storage/vsphere-configurable-maximum-allowed-number-of-block-volumes-per-node.md

```diff
@@ -0,0 +1,200 @@
+By default, a node can have up to four SCSI controllers, each supporting up to 15 devices, allowing for a maximum of 60
+volumes per node (59 + root volume).
+
+However, vSphere version 8 increased the maximum number of volumes per node to 256 (255 + root volume). This enhancement
+aims to leverage this increased limit and provide administrators with finer-grained control over volume allocation
+allowing them to configure the maximum number of block volumes that can be attached to a single node.
+
+Details about configuration maximums: https://configmax.broadcom.com/guest?vmwareproduct=vSphere&release=vSphere%208.0&categories=3-0
+Volume limit configuration for vSphere storage plug-in: https://techdocs.broadcom.com/us/en/vmware-cis/vsphere/container-storage-plugin/3-0/getting-started-with-vmware-vsphere-container-storage-plug-in-3-0/vsphere-container-storage-plug-in-concepts/configuration-maximums-for-vsphere-container-storage-plug-in.html
+
+## Motivation
+
+### User Stories
+
+- As a vSphere administrator, I want to configure the maximum number of volumes that can be attached to a node, so that
+  I can optimize resource utilization and prevent oversubscription.
+- As a cluster administrator, I want to ensure that the vSphere CSI driver operates within the limits imposed by the 
+  underlying vSphere infrastructure.
+
+### Goals
+
+- Provide administrators with granular control over volume allocation on vSphere nodes.
+- Improve resource utilization and prevent oversubscription.
+- Ensure compatibility with existing vSphere infrastructure limitations.
+- Maintain backward compatibility with existing deployments.
+
+### Non-Goals
+
+- Dynamically adjust the limit based on real-time resource usage.
+- Implement per-namespace or per-workload volume limits.
+- Modify the underlying vSphere VM configuration.
+
+## Proposal
+
+1. Enable Feature State Switch (FSS):
+
+   - Use FSS of the vSphere driver to control the activation of the maximum volume limit functionality.
+   - The operator will check for vSphere version 8 (`VCenterChecker`) and conditionally set higher volume limit if version 8 or higher is detected.
+
+2. API for Maximum Volume Limit:
+
+   - Introduce a new field `spec.driverConfig.vSphere.maxAllowedBlockVolumesPerNode` in ClusterCSIDriver API to allow administrators to configure the desired maximum number of volumes per node.
+   - The vSphere CSI operator will read the configured value from the API.
+
+3. Update CSI Node Pods:
+
+   - If the new `maxAllowedBlockVolumesPerNode` API field is set in ClusterCSIDriver the operator will inject the `MAX_VOLUMES_PER_NODE` environment variable into node pods using a DaemonSet hook.
+
+4. Driver Behavior:
+
+   - The vSphere CSI driver will continue to perform basic validation on the user-defined limit, allowing the new limit of 255 volumes per node only on vSphere versions higher than 8.
+   - The driver will respect the configured limit when provisioning volumes.
+
```

Comment:
Add section for documentation changes - we need to document that we don't currently support mixed versions across ESXi hosts in the cluster (all have to be version 8 or higher).

Reply:
Done.